packages/jupyter-ai/jupyter_ai/history.py

```diff
@@ -14,27 +17,53 @@ class BoundedChatHistory(BaseChatMessageHistory, BaseModel):
     messages and 2 AI messages.
     """
 
-    messages: List[BaseMessage] = Field(default_factory=list)
-    size: int = 0
     k: int
+    clear_time: float = 0.0
+    _all_messages: List[BaseMessage] = PrivateAttr(default_factory=list)
+
+    @property
+    def messages(self) -> List[BaseMessage]:
+        return self._all_messages[-self.k * 2 :]
 
     async def aget_messages(self) -> List[BaseMessage]:
         return self.messages
 
     def add_message(self, message: BaseMessage) -> None:
         """Add a self-created message to the store"""
-        self.messages.append(message)
-        self.size += 1
-
-        if self.size > self.k * 2:
-            self.messages.pop(0)
+        self._all_messages.append(message)
 
     async def aadd_messages(self, messages: Sequence[BaseMessage]) -> None:
         """Add messages to the store"""
         self.add_messages(messages)
 
     def clear(self) -> None:
-        self.messages = []
+        self._all_messages = []
+        self.clear_time = time.time()
 
     async def aclear(self) -> None:
         self.clear()
+
+
+class WrappedBoundedChatHistory(BaseChatMessageHistory, BaseModel):
+    """
+    Wrapper around `BoundedChatHistory` to prevent adding messages to the store
+    if a clear was triggered after message started being processed.
+    """
+
+    history: BoundedChatHistory
+    human_msg: HumanChatMessage
```

Comment:
I had to read the implementation a few times to understand why this extra class is necessary. Basically, every time we need to get the chat history object, we want to make sure it only appends an `AgentChatMessage` if the `HumanChatMessage` it is replying to was not cleared. So the memory object needs a reference to the last human message it is replying to, currently via the `human_msg` configuration parameter. 

This makes sense to me, but I think we should do more to make this easier to understand for future devs.

1. Can you rename `human_msg` to `last_human_msg` for clarity? This helps emphasize that this attribute is supposed to refer to the `HumanChatMessage` that a chat handler is replying to (which is always the last human chat message when `get_llm_chat_history()` is called).

2. Can you edit the `WrappedBoundedChatHistory` docstring to the more detailed docstring below (after removing the backslashes)? I think this extra detail is important given the lack of accessible documentation on the details of LangChain memory.

```py
"""
Wrapper around `BoundedChatHistory` that only appends an `AgentChatMessage`
if the `HumanChatMessage` it is replying to was not cleared. If a chat
handler is replying to a `HumanChatMessage`, it should pass this object via
the `last_human_msg` configuration parameter.

For example, a chat handler that is streaming a reply to a
`HumanChatMessage` should be called via:

\```py
async for chunk in self.llm_chain.astream(
    {"input": message.body},
    config={"configurable": {"last_human_msg": message}},
):
    ...
\```

Reference: https://python.langchain.com/v0.1/docs/expression_language/how_to/message_history/
"""
```

Reply:
yea admittedly this wrapped object thing is not a very clean solution but any other solution i could think of was significantly uglier. Can consider refactoring in the future if something nicer is available.

Reply:
Agreed; I wasn't able to come up with a cleaner solution in a short amount of time.